flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/table/TableSourceTest.scala

```diff
@@ -121,19 +161,81 @@ class TableSourceTest extends TableTestBase {
       )
     util.verifyTable(t, expected)
   }
+
+  @Test
+  def testProjectableProcTimeTableSource(): Unit = {
+    // ensures that projection is not pushed into table source with proctime indicators
+    val util = streamTestUtil()
+
+    val projectableTableSource = new TestProctimeSource("pTime") with ProjectableTableSource[Row] {
+      override def projectFields(fields: Array[Int]): TableSource[Row] = {
+        // ensure this method is not called!
+        Assert.fail()
+        null.asInstanceOf[TableSource[Row]]
+      }
+    }
+    util.tableEnv.registerTableSource("PTimeTable", projectableTableSource)
+
+    val t = util.tableEnv.scan("PTimeTable")
+      .select('name, 'val)
+      .where('val > 10)
+
+    val expected =
+      unaryNode(
+        "DataStreamCalc",
+        "StreamTableSourceScan(table=[[PTimeTable]], fields=[id, val, name, pTime])",
+        term("select", "name", "val"),
+        term("where", ">(val, 10)")
+      )
+    util.verifyTable(t, expected)
+  }
+
+  @Test
+  def testProjectableRowTimeTableSource(): Unit = {
```

Comment:
Does filter push down work? With rowtime and proctime?

Reply:
Yes, that should not be a problem. Projection push-down is not possible because the schema of the table is partially constructed inside of the Table API (e.g., appending the proctime attribute).